File: kotlin_server/src/main/kotlin/com/jolimoi_test/kotlin_server/service_impl/services/RomanNumeralNotifierServiceImpl.kt
package com.jolimoi_test.kotlin_server.service_impl.services

import com.jolimoi_test.kotlin_server.repository.NotifierRepository
import org.springframework.http.codec.ServerSentEvent
import org.springframework.messaging.MessageHandler
import org.springframework.messaging.SubscribableChannel
import org.springframework.messaging.support.GenericMessage
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import java.util.*

@Service
class RomanNumeralNotifierServiceImpl(private val channel: SubscribableChannel) : NotifierRepository {
    override fun register(clientId: String): Flux<ServerSentEvent<String>> = Flux.create { sink ->
        val messageHandler = MessageHandler { message ->
            val jobData: String = message.payload as String
            sink.next(
                ServerSentEvent.builder<String>()
                    .id("$clientId-${UUID.randomUUID()}")
                    .event("message")
                    .data(jobData)
                    .build()
            )
        }
        channel.subscribe(messageHandler)
    }

    override suspend fun close(clientId: String): Boolean = channel.unsubscribe { }

    override suspend fun notify(clientId: String, data: String) {
        channel.send(GenericMessage<String>(data))
    }
}

File: node_server/src/api/service_impl/implementations/roman_numeral_notifier_service_impl.ts
import NotifierRepository from "@api/repository/notifier_repository";
import e from "express";
import SSeManager from "../../../domain/sse_manager";

class RomanNumeralNotifierServiceImpl extends NotifierRepository {
    private readonly manager: SSeManager;

    constructor() {
        super();
        this.manager = new SSeManager();
    }

    register(clientId: string, request: e.Request, response: e.Response): void {
        this.manager.addClient(request, response);
    }

    async notify(clientId: string, data: string) {
        await this.manager.send(clientId, data);
    }

    close(): void {
        this.manager.close();
    }
}

const notifierInstance = new RomanNumeralNotifierServiceImpl();
export default notifierInstance;